File: nrworkflow/internal/model/ticket.go
package model

import (
	"database/sql"
	"encoding/json"
	"time"
)

// Status represents the ticket status
type Status string

const (
	StatusOpen       Status = "open"
	StatusInProgress Status = "in_progress"
	StatusClosed     Status = "closed"
)

// IssueType represents the type of ticket
type IssueType string

const (
	IssueTypeBug     IssueType = "bug"
	IssueTypeFeature IssueType = "feature"
	IssueTypeTask    IssueType = "task"
	IssueTypeEpic    IssueType = "epic"
)

// Ticket represents a ticket in the system
type Ticket struct {
	ID          string         `json:"id"`
	ProjectID   string         `json:"project_id"`
	Title       string         `json:"title"`
	Description sql.NullString `json:"-"`
	Status      Status         `json:"status"`
	Priority    int            `json:"priority"`
	IssueType   IssueType      `json:"issue_type"`
	CreatedAt   time.Time      `json:"created_at"`
	UpdatedAt   time.Time      `json:"updated_at"`
	ClosedAt    sql.NullTime   `json:"-"`
	CreatedBy   string         `json:"created_by"`
	CloseReason sql.NullString `json:"-"`
	AgentsState sql.NullString `json:"-"`
}

// MarshalJSON implements custom JSON marshaling for Ticket
func (t Ticket) MarshalJSON() ([]byte, error) {
	// Handle nullable fields
	var description *string
	if t.Description.Valid {
		description = &t.Description.String
	}

	var closedAt *time.Time
	if t.ClosedAt.Valid {
		closedAt = &t.ClosedAt.Time
	}

	var closeReason *string
	if t.CloseReason.Valid {
		closeReason = &t.CloseReason.String
	}

	var agentsState *string
	if t.AgentsState.Valid {
		agentsState = &t.AgentsState.String
	}

	return json.Marshal(&struct {
		ID          string     `json:"id"`
		ProjectID   string     `json:"project_id"`
		Title       string     `json:"title"`
		Description *string    `json:"description"`
		Status      Status     `json:"status"`
		Priority    int        `json:"priority"`
		IssueType   IssueType  `json:"issue_type"`
		CreatedAt   time.Time  `json:"created_at"`
		UpdatedAt   time.Time  `json:"updated_at"`
		ClosedAt    *time.Time `json:"closed_at"`
		CreatedBy   string     `json:"created_by"`
		CloseReason *string    `json:"close_reason"`
		AgentsState *string    `json:"agents_state"`
	}{
		ID:          t.ID,
		ProjectID:   t.ProjectID,
		Title:       t.Title,
		Description: description,
		Status:      t.Status,
		Priority:    t.Priority,
		IssueType:   t.IssueType,
		CreatedAt:   t.CreatedAt,
		UpdatedAt:   t.UpdatedAt,
		ClosedAt:    closedAt,
		CreatedBy:   t.CreatedBy,
		CloseReason: closeReason,
		AgentsState: agentsState,
	})
}

// Dependency represents a dependency between tickets
type Dependency struct {
	ProjectID   string    `json:"project_id"`
	IssueID     string    `json:"issue_id"`
	DependsOnID string    `json:"depends_on_id"`
	Type        string    `json:"type"`
	CreatedAt   time.Time `json:"created_at"`
	CreatedBy   string    `json:"created_by"`
}

File: ui/src/types/ticket.ts
export type Status = 'open' | 'in_progress' | 'closed'
export type IssueType = 'bug' | 'feature' | 'task' | 'epic'

export interface Ticket {
  id: string
  title: string
  description: string | null
  status: Status
  priority: number
  issue_type: IssueType
  parent_ticket_id?: string | null
  created_at: string
  updated_at: string
  closed_at: string | null
  created_by: string
  close_reason: string | null
}

export interface WorkflowProgress {
  workflow_name: string
  current_phase: string
  completed_phases: number
  total_phases: number
  status: string
}

export interface PendingTicket extends Ticket {
  is_blocked: boolean
  blocked_by?: string[]
  workflow_progress?: WorkflowProgress
}

export interface Dependency {
  issue_id: string
  depends_on_id: string
  type: string
  created_at: string
  created_by: string
  depends_on_title?: string
  issue_title?: string
}

export interface TicketWithDeps extends Ticket {
  blockers: Dependency[]
  blocks: Dependency[]
  is_blocked?: boolean
  blocked_by?: string[]
  children?: Ticket[]
  parent_ticket?: Ticket | null
  siblings?: Ticket[]
}

export interface CreateTicketRequest {
  id?: string
  title: string
  description?: string
  priority?: number
  issue_type?: IssueType
  created_by: string
  parent_ticket_id?: string
}

export interface UpdateTicketRequest {
  title?: string
  description?: string
  status?: Status
  priority?: number
  issue_type?: IssueType
  parent_ticket_id?: string
}

export interface TicketListResponse {
  tickets: PendingTicket[]
  total_count: number
  page: number
  per_page: number
  total_pages: number
}

export interface SearchResponse {
  tickets: PendingTicket[]
  query: string
}

export interface DailyStats {
  date: string
  tickets_created: number
  tickets_closed: number
  tokens_spent: number
  agent_time_sec: number
}

export interface StatusResponse {
  counts: {
    open: number
    in_progress: number
    closed: number
    blocked: number
    total: number
  }
  ready_count: number
  pending_tickets: PendingTicket[]
  recent_closed: Ticket[]
}
